web-common/src/features/entity-management/watch-resources-client.ts

```diff
@@ -0,0 +1,92 @@
+import { WatchRequestClient } from "@rilldata/web-common/runtime-client/watch-request-client";
+import {
+  getRuntimeServiceGetResourceQueryKey,
+  getRuntimeServiceListResourcesQueryKey,
+  V1Resource,
+  V1WatchResourcesResponse,
+} from "@rilldata/web-common/runtime-client";
+import { runtime } from "@rilldata/web-common/runtime-client/runtime-store";
+import type { QueryClient } from "@tanstack/svelte-query";
+import { get } from "svelte/store";
+
+export function startWatchResourcesClient(queryClient: QueryClient) {
+  return new WatchRequestClient<V1WatchResourcesResponse>(
+    (runtime) =>
+      `${runtime.host}/v1/instances/${runtime.instanceId}/resources/-/watch`,
+    (res) => handleWatchResourceResponse(queryClient, res),
+    () => invalidateAllResources(queryClient)
+  ).start();
+}
+
+function handleWatchResourceResponse(
+  queryClient: QueryClient,
+  res: V1WatchResourcesResponse
+) {
+  if (!res.resource) return;
+
+  const instanceId = get(runtime).instanceId;
+  // invalidations will wait until the re-fetched query is completed
+  // so, we should not `await` here
+  switch (res.event) {
+    case "RESOURCE_EVENT_ADDED":
+      queryClient.refetchQueries(
+        getRuntimeServiceListResourcesQueryKey(instanceId)
+      );
+    // eslint-disable-next-line no-fallthrough
+    case "RESOURCE_EVENT_UPDATED_SPEC":
+    case "RESOURCE_EVENT_UPDATED_STATE":
+      invalidateResource(queryClient, instanceId, res.resource);
+      break;
+
+    case "RESOURCE_EVENT_DELETED":
```

Comment:
Note that these event names changed slightly here: https://github.com/rilldata/rill/pull/3010